# Snakemake workflow for various types of metagenomic analyses.

include: "rules/common.smk"
include: "rules/other.smk"

include: "rules/annotation.smk"
include: "rules/assembly.smk"
include: "rules/binning.smk"
include: "rules/classification.smk"
include: "rules/preprocessing.smk"
include: "rules/quantification.smk"
include: "rules/taxonomy.smk"

def all_input(wildcards):
    from scripts.common import classify_input, annotation_input, binning_input
    """
    Function defining all requested inputs for the rule all (below).
    """
    wanted_input = []

    if config["run_preprocessing"] or config["fastqc"]:
        wanted_input.append(opj(config["report_path"], "samples_report.html"))

    if config["run_assembly"]:
        wanted_input.append(opj(config["report_path"], "assembly",
                                "assembly_stats.pdf"))
        wanted_input += annotation_input(config, assemblies)
        wanted_input += binning_input(config, assemblies)

    wanted_input += classify_input(config)

    return wanted_input

##### master target rule #####
rule all:
    input: all_input
